CSS Nesting: implement CSSOM (insertRule, deleteRule, cssRules)

mdubet opened the pull request:
#### 674ff8500b9cc3ad12d52ca88f5e64f9c28ceba0
<pre>
CSS Nesting: implement CSSOM (insertRule, deleteRule, cssRules)
<a href="https://bugs.webkit.org/show_bug.cgi?id=251374">https://bugs.webkit.org/show_bug.cgi?id=251374</a>
rdar://103844456

Reviewed by NOBODY (OOPS!).

StyleRule now inherits from GroupingRule (like @media, @supports...)
It allows us to share all of the CSSOM code between the various CSS rules
which support grouping.

The StyleRule IDL file exposes the insertRule/deleteRule/cssRules methods
directly (and not through inheritance from GroupingRule IDL) to allow for
activation/desactivation by the CSSNesting feature flag.

The serialization via cssText() is still not exactly the one expected by WPT,
as it&apos;s currently still un-speced.

Some of the WPT changes are import, the rest will be exported.

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalid-inner-rules-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/serialize-group-rules-with-decls.tentative-expected.txt:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::insertRule):
(WebCore::CSSGroupingRule::deleteRule):
(WebCore::CSSGroupingRule::cssTextForDeclsAndRules const):
* Source/WebCore/css/CSSRule.cpp:
(WebCore::CSSRule::hasStyleRuleAscendant const):
* Source/WebCore/css/CSSRule.h:
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::isNestingSelector const):
* Source/WebCore/css/CSSSelectorList.h:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::CSSStyleRule):
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::cssText const):
(WebCore::CSSStyleRule::length const): Deleted.
(WebCore::CSSStyleRule::item const): Deleted.
(WebCore::CSSStyleRule::cssRules const): Deleted.
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSStyleRule.idl:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::averageSizeInBytes):
(WebCore::StyleRule::StyleRule):
(WebCore::StyleRule::create):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::StyleRuleBase):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseRule):
(WebCore::CSSParser::parseSelector):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::parseRule):
(WebCore::CSSParserImpl::createNestingParentRule):
(WebCore::CSSParserImpl::consumeStyleRule):
* Source/WebCore/css/parser/CSSParserImpl.h:
(WebCore::CSSParserImpl::isNestedContext):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::parseCSSSelector):
(WebCore::CSSSelectorParser::CSSSelectorParser):
(WebCore::CSSSelectorParser::consumeNestedSelectorList):
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::isValidRuleHeaderText):
(WebCore::isNestedContext):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addStyleRule):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3650a9c20560e608f4eddef9173771a36514ec6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/777 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/801 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/830 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1045 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/693 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/768 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/849 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/884 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/918 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/785 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/748 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/740 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1034 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/804 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/732 "6 failures") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/762 "2 flakes 143 failures") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/707 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/755 "1 flakes 8 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1748 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/740 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/721 "3 failures") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/691 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/740 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/750 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

anttijk reviewed (changes requested):
This patch is too big. Can you make one without all the random cssText refactoring for example?

mdubet committed:
CSS Nesting: implement CSSOM (insertRule, deleteRule, cssRules)
https://bugs.webkit.org/show_bug.cgi?id=251374
rdar://103844456

Reviewed by NOBODY (OOPS!).

The CSSOM now only works with StyleRuleWithNesting, losing the StyleRule
memory optimization.

The StyleRule IDL file exposes the insertRule/deleteRule/cssRules methods
directly (and not through inheritance from GroupingRule IDL) to allow for
activation/desactivation by the CSSNesting feature flag.

The serialization via cssText() is still not exactly the one expected by WPT,
as it's not completely specified.

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalid-inner-rules-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/serialize-group-rules-with-decls.tentative-expected.txt:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::insertRule):
(WebCore::CSSGroupingRule::deleteRule):
(WebCore::CSSGroupingRule::cssTextForDeclsAndRules const):
* Source/WebCore/css/CSSRule.cpp:
(WebCore::CSSRule::hasStyleRuleAscendant const):
* Source/WebCore/css/CSSRule.h:
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::isNestingSelector const):
* Source/WebCore/css/CSSSelectorList.h:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::CSSStyleRule):
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::cssText const):
(WebCore::CSSStyleRule::length const): Deleted.
(WebCore::CSSStyleRule::item const): Deleted.
(WebCore::CSSStyleRule::cssRules const): Deleted.
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSStyleRule.idl:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::averageSizeInBytes):
(WebCore::StyleRule::StyleRule):
(WebCore::StyleRule::create):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::StyleRuleBase):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseRule):
(WebCore::CSSParser::parseSelector):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::parseRule):
(WebCore::CSSParserImpl::createNestingParentRule):
(WebCore::CSSParserImpl::consumeStyleRule):
* Source/WebCore/css/parser/CSSParserImpl.h:
(WebCore::CSSParserImpl::isNestedContext):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::parseCSSSelector):
(WebCore::CSSSelectorParser::CSSSelectorParser):
(WebCore::CSSSelectorParser::consumeNestedSelectorList):
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::isValidRuleHeaderText):
(WebCore::isNestedContext):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addStyleRule):

mdubet committed:
StyleRuleGroup should use composition instead of inheritance
https://bugs.webkit.org/show_bug.cgi?id=254297
rdar://107103847

Reviewed by NOBODY (OOPS!).

The inheritance comes from the CSSGroupingRule spec,
but we have no reason to follow the same pattern/hierarchy in our
internal representation.

* Source/WebCore/css/CSSConditionRule.cpp:
(WebCore::CSSConditionRule::CSSConditionRule):
* Source/WebCore/css/CSSConditionRule.h:
* Source/WebCore/css/CSSContainerRule.cpp:
(WebCore::CSSContainerRule::styleRuleContainer const):
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::CSSGroupingRule):
(WebCore::CSSGroupingRule::~CSSGroupingRule):
(WebCore::CSSGroupingRule::insertRule):
(WebCore::CSSGroupingRule::deleteRule):
(WebCore::CSSGroupingRule::cssTextForDeclsAndRules const):
(WebCore::CSSGroupingRule::length const):
(WebCore::CSSGroupingRule::item const):
(WebCore::CSSGroupingRule::reattach):
(WebCore::CSSGroupingRule::groupRule const):
(WebCore::CSSGroupingRule::groupRule):
* Source/WebCore/css/CSSGroupingRule.h:
(WebCore::CSSGroupingRule::rule const):
(WebCore::CSSGroupingRule::groupRule const): Deleted.
(WebCore::CSSGroupingRule::groupRule): Deleted.
* Source/WebCore/css/CSSLayerBlockRule.cpp:
(WebCore::CSSLayerBlockRule::name const):
* Source/WebCore/css/CSSMediaRule.cpp:
(WebCore::CSSMediaRule::mediaQueries const):
(WebCore::CSSMediaRule::setMediaQueries):
* Source/WebCore/css/CSSSupportsRule.cpp:
(WebCore::CSSSupportsRule::conditionText const):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleGroup::StyleRuleGroup):
(WebCore::StyleRuleGroup::childRules):
(WebCore::StyleRuleGroup::fromStyleRuleBase):
(WebCore::StyleRuleMedia::StyleRuleMedia):
(WebCore::StyleRuleSupports::StyleRuleSupports):
(WebCore::StyleRuleLayer::StyleRuleLayer):
(WebCore::StyleRuleContainer::StyleRuleContainer):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isStyleRuleWithNesting const):
(WebCore::StyleRuleBase::isGroupRule const):
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::traverseRulesInVector):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::shouldDirtyAllStyle):

anttijk commented on Source/WebCore/css/CSSStyleRule.h:
```diff
@@ -34,9 +34,8 @@ class StyleRuleCSSStyleDeclaration;
 class StyleRuleWithNesting;
 
-class CSSStyleRule final : public CSSRule, public CanMakeWeakPtr<CSSStyleRule> {
+class CSSStyleRule final : public CSSGroupingRule, public CanMakeWeakPtr<CSSStyleRule> {
```
I think you are just making problems for yourself with this. This leads to all the StyleRuleGroup refactoring. It also doesn't match the structure of the idl interface.

You could find other ways to share the child management code, or just copy it for now.

patrickangle commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1052,6 +1051,7 @@ void InspectorStyleSheet::reparseStyleSheet(const String& text)
     {
         CSSStyleSheet::RuleMutationScope mutationScope(m_pageStyleSheet.get());
         m_pageStyleSheet->contents().parseString(text);
+        m_pageStyleSheet->contents().transformStyleRuleToNesting();
```
Is there ever a time we wouldn't want to do this after parsing sheet text? If not, why wouldn't calling `parseString` put the contents into the correct state itself?

mdubet replied:
Yes indeed, it should be handled somewhere in the callee code, it's a leftover
